[FIX] product,(website_)sale: fix product configurator performance

seb-odoo opened the pull request:
Merge commit for the PR #30274, improving performance of product configurator.

- fix compute of list view of variants
- create generator to iterate the possible combinations
- avoid to compute useless information/combinations
- avoid to compute useless image resize
- take advantage of cache and prefetch (need to be improved in master with
	real fields and correct dependencies)
- add some missing index
- add some tests
- ...

task-1922055
opw-1923999

seb-odoo commented:
@JKE-be quick and easy PR, this does not solve all the slowness (I keep investigating), but it does improve it by a lot when in the wrong conditions.

Either we can merge this one directly, or we wait until I find more to merge all of them.

sswapnesh commented on addons/website_sale/models/product.py:
```diff
@@ -179,7 +179,10 @@ def _is_quick_add_to_cart_possible(self, parent_combination=None):
+        # Ideally the following condition should test for _get_possible_variants
+        # but it is too slow to call for multiple products, so we just check
+        # for all the variants (possible or not).
+        if len(self.product_variant_ids) != 1:
```
Better

- if len(self.`with_prefetch()`.product_variant_ids) != 1 


(Not sure if `>` will be faster or not instaed of `!=` but `>` is used in base already for similar condition)

- or we can use field `product_variant_count` 


https://github.com/odoo/odoo/blob/12.0/addons/product/models/product_template.py#L123


What do you think on this ?

seb-odoo replied:
It is actually still a problem (this and also my suggestion solution) because after thinking about it we shouldn't try to read all of the variants, just know if there are at least 2 possible.

I'm currently trying to update _get_first_possible_combination into _get_nth_possible_combination, so use it as a way to know if there are at least 2 without having to read any of them.

sswapnesh replied:
Will  `with_context(prefetch_fields=False)`   make sense here ? (Not sure about this)

seb-odoo replied:
I appreciate the feedback. I think you are correct that disabling `prefetch` would be good if I wanted to keep using `product_variant_ids` for this, or even using `product_variant_count`.

But I think it is better to keep working with the combinations instead of the variants like I did for the rest of the configurator fixes, and so in priority I want to try to not use `product_variant_ids` if I find a way to not use it.

sswapnesh commented on addons/product/models/product_attribute.py:
```diff
@@ -246,9 +246,9 @@ class ProductTemplateAttributeExclusion(models.Model):
 
     product_template_attribute_value_id = fields.Many2one(
-        'product.template.attribute.value', string="Attribute Value", ondelete='cascade')
+        'product.template.attribute.value', string="Attribute Value", ondelete='cascade', index=True)
```
Is is acceptable to create Index for existing field in stable version ?

Yenthe666 replied:
I believe it is OK as long as there is a good reason (big perf impact) and we are not very far into the release. Not 100 percent sure though.

JKE-be commented:
Yes you can but it will be only apply in new database or after update

seb-odoo commented:
Runbot green now, all good for me.

For the record, on a simple demo database (without 1000 variants per product), the slowest part of rendering /shop after this PR is `_compute_images` at around 20% of the time, when `_get_variant_for_combination` is "only" around 10%.

seb-odoo commented:
I think I found a correct way to cache the result of the most problematic method. The first call on each worker is still slow (but much better than before for sure), and once the result is cached this PR should reduce the time of /shop by 40% on runbot with our demo data (a bit above 1s, from 2s before), and reduce much more than that for worse case scenario (tons of variants per product, massive images to resize, ...).

seb-odoo commented:
Runbot now green, and I see no more obvious improvement to do. We can start to review.

The typical demo case on runbot (after cache) went from 1930ms to 1268ms. This is around 35% gain on a relatively simple case.
Worse case could have been 20s+ before, also down to around 1s now.

The first call on each worker before cache is just a bit better than before but it's not possible to solve further without changing models.

seb-odoo commented:
@JKE-be reminder here too

JKE-be commented:
@robodoo delegate=seb-odoo

tu sais mieux que moi ce que ca fait et comment on fait!

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -824,16 +831,24 @@ def _get_variant_for_combination(self, combination):
+        return self._get_nth_possible_combination(parent_combination, necessary_values, n=1)
+
+    @api.multi
+    def _get_nth_possible_combination(self, parent_combination=None, necessary_values=None, n=1):
```
don't you have a function with an iterator that return each combinaisons?
If not, probably a good thing to do, and just iters 2 times to check if more than one.

seb-odoo replied:
Done, 1st commit updated by using generator instead of get_nth. This doesn't have a significant performance impact, but that code is more understandable this way (and not more nonlocal).

seb-odoo commented:
@JKE-be Can we make a final review soon?

We just tested the task locally for opw-1923999 with SBB and it does improve the situation as expected! From 2s+ down to 700ms after cache.

seb-odoo commented:
Rebased on latest 12.0, updated commit dates, added ref to new opw, and runbot green. Ready to go for me.

lmignon commented on addons/product/models/product.py:
```diff
@@ -338,6 +338,8 @@ def create(self, vals_list):
             if not (self.env.context.get('create_from_tmpl') and len(product.product_tmpl_id.product_variant_ids) == 1):
                 product._set_standard_price(vals.get('standard_price') or 0.0)
+        self.invalidate_cache()
+        self.clear_caches()
```
Why do you invalidate the cache here? Is this the actual record's cache or the overall cache? If the global cache is invalidated here, I'm pretty sure you are introducing others performance issue in others contexts.  (same comment applies for 'write' and 'unlink'

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -848,63 +862,81 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
 
         if not self.active:
-            return self.env['product.template.attribute.value']
+            return _("The product template is archived so no combination is possible.")
```
return Falsy value (False or empty browse_Record) instead of string
idem below

seb-odoo replied:
Ok, I can just return nothing then. The idea was to give the user a better message if that exception is raised, but I guess we can consider it a dev error if that happens and not worry about it?

JKE-be replied:
as discuss, since string will be the exception message, => we keep it

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -824,22 +831,29 @@ def _get_variant_for_combination(self, combination):
+        return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
+
+    @api.multi
+    def _get_possible_combinations(self, parent_combination=None, necessary_values=None):
```
necessary_values=[] and remove fallback later

seb-odoo replied:
I'm fine either way but we've had issues before when swapping recordset with []. It's obviously fine when only iterating which is the case here though, but that's why I preferred to only fallback in the iteration specifically.

seb-odoo commented:
@lmignon 

The problem is that `_get_variant_for_combination` is slow by design but we can't change the model in stable. So in order to improve the situation, its result is cached. However, the complete dependency tree to compute that result it way too complex to express and also too costly to compute just to know if we should invalidate the cached result or not.

So I took the opposite approach of invalidating the cache whenever I knew it had to be done. Unfortunately it is a global cache indeed, but fortunately I only have to do it in very specific cases that happen only during set up and configuration. It's not every minute that someone has to create a new product or resequence attributes for example.

If you have a better solution, or if it is possible to invalidate the cache only for specific models, I'm all for it of course, but I have not found such a possibility without having to adapt ORM methods. But if it turns out to really be a problem maybe that can be done, I just don't think it's necessary.

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -848,63 +862,81 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
-                that have to be tested for possibility
+        # keep going while we have attribute lines to test
+        while len(ptal_stack):
+            attribute_lines = ptal_stack.pop()
```
atrribute_line = no?

and useless [0] later

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -848,63 +862,81 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
+            return _("The product template is archived so no combination is possible.")
 
-        attribute_lines = self._get_valid_product_template_attribute_lines()
+        ptal_stack = [self._get_valid_product_template_attribute_lines()]
```
recordset are iterable by default

seb-odoo replied:
This is a stack of recordset (with potentially multiple values in each).

[0] = all attributes lines
[1] = all attributes lines - 1
[2] = all attributes lines - 2
etc until it's empty which is the stop condition for the loop.

The combination_stack is doing the reverse.
[0] = no value
[1] = value 2 of line n-1
[2] = value 1 of line n-1
etc

lmignon commented:
Thank you for your prompt reply @seb-odoo 
> It's not every minute that someone has to create a new product or resequence attributes for example.

It's a common UC for a lot of our customers to use the import/export functionality to keep the products catalog up to date (saisonal products, information provided by the product's suppliers / manufacturers, ...) I've the feeling that your changes could make this process unusable if the files contains a mix of new products and products to deactivate... Do you have tested the impact of your changes if you create 1000 new products at once by importing a csv file? The same kind of test should be don if you deactivate a lot of products at once. We must be sure that this kind of change doesn't introduce regression in terms of performance for customer managing thousands of products into Odoo. If it's the case, for sure our curstomers will not be happy at all and we'll introduce a new issue to ask a solution for the regression in terms of perf.... 
I may be frightened a little too quickly, but I think this type of change needs to be accompanied by extensive testing to make sure it doesn't degrade the user experience.

JKE-be commented on addons/website_sale/models/product.py:
```diff
@@ -179,17 +182,18 @@ def _is_quick_add_to_cart_possible(self, parent_combination=None):
             return False
-        return True
+        except StopIteration:
+            if self._has_is_custom_values():
```
```python
except YourException:
   pass

if self...
```

JKE-be replied:
btw

in your case, you can give a default value a second parameter of next to avoid this StopIteration

seb-odoo replied:
Same as above, I just need to test if the iteration works here, I don't want to return its result.

seb-odoo replied:
Following our discussion, using `next(gen, False) is not False` instead

JKE-be commented on addons/sale/models/product_template.py:
```diff
@@ -290,9 +290,13 @@ def _is_add_to_cart_possible(self, parent_combination=None):
+            # for performance: avoid calling `_get_possible_combinations`
+            return False
+        try:
+            next(self._get_possible_combinations(parent_combination))
```
next(iterator[, default])
> Retrieve the next item from the iterator by calling its __next__() method. If **default** is given, it is returned if the iterator is exhausted, otherwise StopIteration is raised.

seb-odoo replied:
I don't want to return the return of the iteration here. The result might be an empty combination that is falsy.  I need to return True specifically.

JKE-be replied:
I dont suggest to return next(xxx,)

but just replace the try except by 

if next(xxx, default):
   return what you want

else continue like before

seb-odoo replied:
Following our discussion, using `next(gen, False) is not False` instead

JKE-be commented on addons/product/models/product_attribute.py:
```diff
@@ -78,6 +82,14 @@ def name_get(self):
+        res = super(ProductAttributeValue, self).write(values)
+        if 'sequence' in values:
+            self.invalidate_cache()
+            self.clear_caches()
```
to check

JKE-be commented on addons/product/models/product_pricelist.py:
```diff
@@ -491,3 +491,11 @@ def _onchange_compute_price(self):
+    @api.multi
+    def write(self, values):
+        res = super(PricelistItem, self).write(values)
+        if 'percent_price' in values:
```
to dblecheck

seb-odoo commented:
I doubled checked all the cache invalidation.

I removed those that were not necessary. For those that I kept, I added a comment to explain why it was needed, and I made sure there was a test that would cover it (test crashing when removing the cache invalidation).

Good to go now?

seb-odoo commented:
Runbot green

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -539,42 +558,68 @@ def _get_valid_product_template_attribute_lines(self):
+    def _compute_valid_existing_variant_ids(self):
+        """This compute is done outside of `_compute_valid_attributes` because
+        it is often not needed, and it can be very bad on performance."""
+        existing_variants = self.env['product.product'].search([('product_tmpl_id', 'in', self.ids), ('active', '=', True)])
```
active=true ? it is by default :/

seb-odoo replied:
If someone put `active_test` in the context, I still want this method to make `active=True`, because that's what makes the field `valid_existing_variant_ids` different than `valid_archived_variant_ids`.

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -539,42 +558,68 @@ def _get_valid_product_template_attribute_lines(self):
+        archived_variants = self.env['product.product'].search([('product_tmpl_id', 'in', self.ids), ('active', '=', False)])
+
+        # prefetch
+        self.mapped('attribute_line_ids.value_ids.id')
```
can you confirm your 

(14:34:33) seb@odoo.com: il y a peut être une autre solution que le map pour faire un prefetch si le map est trop lent? j'ai pas besoin de résultat du map là
(14:35:06) seb@odoo.com: genre self.attribute_line_ids[:1].value_ids.id 
(14:35:08) seb@odoo.com: à tester

JKE-be replied:
(16:26:08) seb@odoo.com: il sera de toute facon genre 10 et pas 1000 ça c'est sur

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -539,42 +558,68 @@ def _get_valid_product_template_attribute_lines(self):
+        """deprecated, use `valid_product_attribute_value_ids`"""
         self.ensure_one()
-        return self._get_valid_product_template_attribute_lines().mapped('value_ids')
+        return self.valid_product_attribute_value_ids
```
add logger warning

seb-odoo replied:
I didn't actually change the existing code using the deprecated methods to keep a diff minimal.

It's not a big deal to use it, since the method is now using the computed field, it's performant too, it's just a duplicate.

JKE-be commented on addons/product/models/product_template.py:
```diff
@@ -839,21 +843,30 @@ def _get_variant_for_combination(self, combination):
-            lambda v: v.product_template_attribute_value_ids == filtered_combination
-        )[:1]
+            lambda v: v.attribute_value_ids == attribute_values
+        )[:1].id
```
in stable :/

JKE-be replied:
here!

JKE-be replied:
new func

seb-odoo commented:
I just rebased on latest 12.0, no changes made following the last comments.

seb-odoo commented:
Updated PR to include code from #30468 and to make further improvements by using only template data on /shop.

Current load time based on demo data can now be < 300ms locally.

Waiting on runbot green + time result.

JKE-be commented on addons/product/models/product.py:
```diff
@@ -279,7 +279,16 @@ def _set_image_value(self, value):
+        # expects both images to be the same.
+        variant_count = self.env['product.product'].search_count([
+            ('product_tmpl_id', '=', self.product_tmpl_id.id),
+            ('active', '=', True),
```
if variant not activated, no archived, so not need to check the active

seb-odoo replied:
Actually there is a field for that: product_variant_count so I'll use it instead.

JKE-be commented:
@robodoo r+ rebase-ff 

Continuity of PR #28079 :dizzy_face:

It is amazing the number of diff when you know that Product Configurator don't change anything for the front end part. What explain why we were able merge all this code in product without check anything else in website. :bowtie:

Hope that 5 months after release, we will be able to start to work on new features... and stop to change template and all logic each time. :v:

robodoo commented:
Merge method set to rebase and fast-forward

JKE-be commented:
or if not to late @robodoo , you can make a rebase-merge

ty

robodoo commented:
Staging failed: ci/runbot (view more at http://runbot.odoo.com/runbot/build/474788)

KangOl commented:
@robodoo retry

robodoo commented:
Merged, thanks!
